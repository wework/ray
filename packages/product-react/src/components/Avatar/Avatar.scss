$avatar-background-color: #e6e6e6;
$avatar-fill-color: #b3b3b3;
$avatar-iconSize: (
  small: 32px,
  medium: 40px,
  large: 148px
);

@mixin avatar-size($size) {
  height: map-get($avatar-iconSize, $size);
  width: map-get($avatar-iconSize, $size);
}

.lucia-avatar {
  border-radius: 50%;
  display: inline-block;
  position: relative;

  img {
    border-radius: 50%;
    max-height: 100%;
    max-width: 100%;
  }

  span {
    float: left;
    left: 50%;
    position: relative;
    text-transform: uppercase;
    top: 50%;
    transform: translate(-50%, -50%);
  }

  &--small {
    @include avatar-size(small);
  }

  &--medium {
    @include avatar-size(medium);
  }

  &--large {
    @include avatar-size(large);
  }
}

.lucia-no-avatar {
  background-color: $avatar-background-color;
}

.lucia-avatar--action {
  cursor: pointer;
  height: 100%;
  position: absolute;
  width: 100%;

  svg {
    background-color: $avatar-fill-color;
    border-radius: 50%;
    bottom: 0;
    fill: $avatar-fill-color;
    height: 48px;
    padding: 8px;
    position: absolute;
    right: 0;
    width: 48px;
  }
}
