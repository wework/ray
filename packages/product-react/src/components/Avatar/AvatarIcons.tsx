import React from 'react';

const AvatarEditIcon = (
  <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
    <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
      <g id="ico-bin" fill="#FFFFFF" fillOpacity="0">
        <rect id="bin" x="0" y="0" width="24" height="24" />
      </g>
      <rect
        id="Rectangle-path"
        fill="#FFFFFF"
        fillRule="nonzero"
        x="7"
        y="21"
        width="13"
        height="1"
      />
      <rect
        id="Rectangle-path"
        fill="#FFFFFF"
        fillRule="nonzero"
        x="4"
        y="21"
        width="2"
        height="1"
      />
      <path
        d="M7,20 L7.5,20 C7.56594566,19.9997888 7.63119813,19.9865344 7.692,19.961 C7.72834515,19.9422957 7.76198516,19.9187476 7.792,19.891 C7.81227353,19.8802064 7.83166667,19.867835 7.85,19.854 L19.85,7.854 C19.9440417,7.76019574 19.9968938,7.63282726 19.9968938,7.5 C19.9968938,7.36717274 19.9440417,7.23980426 19.85,7.146 L16.85,4.146 C16.7561957,4.0519583 16.6288273,3.99910622 16.496,3.99910622 C16.3631727,3.99910622 16.2358043,4.0519583 16.142,4.146 L4.142,16.146 C4.12940941,16.163384 4.11772711,16.1814082 4.107,16.2 C4.07848061,16.2327731 4.05428313,16.2690694 4.035,16.308 C4.01083484,16.3690797 3.99894165,16.4343223 3.99993509,16.5 L3.99993509,20 L7,20 Z M16.5,5.207 L18.793,7.5 L17.5,8.793 L15.207,6.5 L16.5,5.207 Z M5,18.5 L5,16.707 L14.5,7.207 L16.793,9.5 L7.293,19 L5,19 L5,18.5 Z"
        id="Shape"
        fill="#FFFFFF"
        fillRule="nonzero"
      />
    </g>
  </svg>
);

const AvatarAddIcon = (
  <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
    <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
      <g id="ico-bin" fill="#FFFFFF" fillOpacity="0">
        <rect id="bin" x="0" y="0" width="24" height="24" />
      </g>
      <path
        d="M12.5,11.5 L20.5,11.5 C20.7761424,11.5 21,11.7238576 21,12 C21,12.2761424 20.7761424,12.5 20.5,12.5 L12.5,12.5 L12.5,20.5 C12.5,20.7761424 12.2761424,21 12,21 C11.7238576,21 11.5,20.7761424 11.5,20.5 L11.5,12.5 L3.5,12.5 C3.22385763,12.5 3,12.2761424 3,12 C3,11.7238576 3.22385763,11.5 3.5,11.5 L11.5,11.5 L11.5,3.5 C11.5,3.22385763 11.7238576,3 12,3 C12.2761424,3 12.5,3.22385763 12.5,3.5 L12.5,11.5 Z"
        id="Combined-Shape"
        fill="#FFFFFF"
        fillRule="nonzero"
      />
    </g>
  </svg>
);

export { AvatarAddIcon, AvatarEditIcon };
