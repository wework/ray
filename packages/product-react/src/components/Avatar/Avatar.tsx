import clsx from 'clsx';
import React, { FC } from 'react';
import './Avatar.scss';
import { AvatarAddIcon, AvatarEditIcon } from './AvatarIcons';

export enum SizeVariations {
  small = 'small',
  medium = 'medium',
  large = 'large'
}

export interface AvatarProps {
  add?: boolean;
  alt?: string;
  edit?: boolean;
  name?: string;
  image?: string;
  size?: keyof typeof SizeVariations;
  className?: string;
  onAddImageClick?: (event?: React.SyntheticEvent<EventTarget>) => void;
}

export const Avatar: FC<AvatarProps> = ({
  add,
  alt,
  edit,
  image,
  name,
  size,
  onAddImageClick,
  className,
  ...rest
}) => {
  const wrapperClass = clsx(
    'lucia-avatar',
    {
      'lucia-avatar--small': size === 'small',
      'lucia-avatar--medium': size === 'medium',
      'lucia-avatar--large': size === 'large',
      'lucia-no-avatar': image === undefined
    },
    className
  );

  const getInitialFontSize = (size: any) => {
    switch (size) {
      case SizeVariations.small:
        return 'ray-text--body-x-small';
      case SizeVariations.large:
        return 'ray-text--body-large';
      default:
        return 'ray-text--body';
    }
  };

  const onAddImageClickAction = (
    event: React.SyntheticEvent<EventTarget>
  ): void => {
    onAddImageClick && onAddImageClick();
  };
  return (
    <div className={wrapperClass} {...rest}>
      {size === SizeVariations.large && (
        <div className="lucia-avatar--action" onClick={onAddImageClickAction}>
          {add && AvatarAddIcon}
          {edit && AvatarEditIcon}
        </div>
      )}
      {image ? (
        <img src={image} alt={alt} />
      ) : (
        name && <span className={getInitialFontSize(size)}>{name[0]} </span>
      )}
    </div>
  );
};

Avatar.defaultProps = {
  add: false,
  edit: false,
  size: SizeVariations.medium
};

export default Avatar;
