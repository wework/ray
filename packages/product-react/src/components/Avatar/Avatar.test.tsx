import Enzyme, { configure, mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { Avatar } from './Avatar';

configure({ adapter: new Adapter() });

describe('Avatar', () => {
  let component: ReactWrapper<any>;
  const fauxCallback = jest.fn();

  beforeEach(() => {
    component = mount(<Avatar />);
  });

  test('should match snapshot and styles', () => {
    expect(mount(<Avatar />)).toMatchSnapshot();
  });

  describe('Default Properties', () => {
    it('has default properties "add" that can be set', () => {
      expect(component.props().add).toEqual(false);
    });
    it('has default properties "edit" that can be set', () => {
      expect(component.props().edit).toEqual(false);
    });
    it('has default properties "size" that can be set', () => {
      expect(component.props().size).toEqual('medium');
    });
    it('has no defaulted image', () => {
      expect(component.props().image).toEqual(undefined);
      expect(
        component
          .find('img')
          .at(0)
          .exists()
      ).toEqual(false);
    });
    it('has no default "add/edit" icon', () => {
      expect(component.props().add).toEqual(false);
      expect(component.props().edit).toEqual(false);
      expect(
        component
          .find('.lucia-avatar--action')
          .at(0)
          .exists()
      ).toEqual(false);
    });
    it('has initial of name when specified', () => {
      component.setProps({
        name: 'Patrick Stewart'
      });
      expect(component.find('img').exists()).toEqual(false);
      expect(
        component
          .find('span')
          .at(0)
          .contains('P')
      ).toEqual(true);
    });
    it('defaults to display image avatar if both image and name is specified', () => {
      component.setProps({
        image:
          'https://pbs.twimg.com/profile_images/624249118114881536/qxn_I_oR_400x400.jpg',
        name: 'Patrick Stewart'
      });
      expect(
        component
          .find('span')
          .at(0)
          .contains('P')
      ).toEqual(false);
      expect(component.find('img').exists()).toEqual(true);
    });
  });

  describe('Classes', () => {
    it('contains DEFAULT class', () => {
      expect(
        component
          .find('.lucia-avatar')
          .at(0)
          .exists()
      ).toEqual(true);
    });
    it('contains nested "add" icon (ONLY for size="large")', () => {
      expect(
        component
          .find('.lucia-avatar--action')
          .at(0)
          .exists()
      ).toEqual(false);
      component.setProps({ add: true });
      expect(
        component
          .find('.lucia-avatar--action')
          .at(0)
          .exists()
      ).toEqual(false);
      component.setProps({ add: true, size: 'small' });
      expect(
        component
          .find('.lucia-avatar--action')
          .at(0)
          .exists()
      ).toEqual(false);
      component.setProps({ add: true, size: 'large' });
      expect(
        component
          .find('.lucia-avatar--action')
          .at(0)
          .exists()
      ).toEqual(true);
    });
    it('contains nested image', () => {
      expect(
        component
          .find('.lucia-avatar--action')
          .at(0)
          .exists()
      ).toEqual(false);
      component.setProps({
        image:
          'https://pbs.twimg.com/profile_images/624249118114881536/qxn_I_oR_400x400.jpg'
      });
      expect(
        component
          .find('img')
          .at(0)
          .exists()
      ).toEqual(true);
    });

    it(' has empty image avatar with name props and size small', () => {
      component.setProps({
        image: '',
        name: 'Patrick Stewart',
        size: 'small'
      });
      expect(
        component
          .find('span')
          .at(0)
          .contains('P')
      ).toEqual(true);
      expect(component.find('img').exists()).toEqual(false);
    });
    it(' has empty image avatar with name props and size large', () => {
      component.setProps({
        image: '',
        name: 'Patrick Stewart',
        size: 'large'
      });
      expect(
        component
          .find('span')
          .at(0)
          .contains('P')
      ).toEqual(true);
      expect(component.find('img').exists()).toEqual(false);
    });

    it(' has broken image avatar with name props and size large', () => {
      component.setProps({
        image:
          'https://pbs.22222222twimg.com/profile_images/624249118114881536/qxn_I_oR_400x400.jpg',
        name: 'Patrick Stewart',
        size: 'large'
      });
      expect(
        component
          .find('span')
          .at(0)
          .contains('P')
      ).toEqual(false);
      expect(component.find('img').exists()).toEqual(true);
    });

    it('contains correct class for "size" variant ', () => {
      expect(
        component
          .find('.lucia-avatar--small')
          .at(0)
          .exists()
      ).toEqual(false);
      component.setProps({ size: 'small' });
      expect(component.find('.lucia-avatar--small').exists()).toEqual(true);
      component.setProps({ size: 'medium' });
      expect(component.find('.lucia-avatar--medium').exists()).toEqual(true);
      component.setProps({ size: 'large' });
      expect(component.find('.lucia-avatar--large').exists()).toEqual(true);
      component.setProps({ size: 'gobblygook' });
      expect(component.find('.lucia-avatar--small').exists()).toEqual(false);
      expect(component.find('.lucia-avatar--medium').exists()).toEqual(false);
      expect(component.find('.lucia-avatar--large').exists()).toEqual(false);
    });
  });

  describe('Functions', () => {
    it('clicks (ONLY for size="large" w/ "add" icon and passed action)', () => {
      component.setProps({
        onAddImageClick: fauxCallback,
        add: true,
        size: 'large'
      });
      component.find('.lucia-avatar--action').simulate('click');
      expect(fauxCallback.mock.calls.length).toEqual(1);
    });
  });
});
