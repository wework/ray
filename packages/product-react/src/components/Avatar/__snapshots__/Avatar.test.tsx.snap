// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should match snapshot and styles 1`] = `
<Avatar
  add={false}
  edit={false}
  size="medium"
>
  <div
    className="lucia-avatar lucia-avatar--medium lucia-no-avatar"
  />
</Avatar>
`;
